Sources/SwiftFormatRules/TypeNamesShouldBeCapitalized.swift

```diff
@@ -0,0 +1,52 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftFormatCore
+import SwiftSyntax
+
+/// `struct`, `class`, `enum` and `protocol` declarations should have a capitalized name.
+///
+/// Lint:  Types with un-capitalized names will yield a lint error.
+public final class TypeNamesShouldBeCapitalized : SyntaxLintRule {
+  public override func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
+    diagnoseNameConventionMismatch(node, name: node.name)
+    return .visitChildren
+  }
+
+  public override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+    diagnoseNameConventionMismatch(node, name: node.name)
+    return .visitChildren
+  }
+
+  public override func visit(_ node: EnumDeclSyntax) -> SyntaxVisitorContinueKind {
+    diagnoseNameConventionMismatch(node, name: node.name)
+    return .visitChildren
+  }
+
+  public override func visit(_ node: ProtocolDeclSyntax) -> SyntaxVisitorContinueKind {
+    diagnoseNameConventionMismatch(node, name: node.name)
+    return .visitChildren
+  }
+
+  private func diagnoseNameConventionMismatch<T: DeclSyntaxProtocol>(_ type: T, name: TokenSyntax) {
+    if let firstChar = name.text.first, !firstChar.isUppercase {
```

Comment:
Another thought occurred to me later; this will also flag type names that start with underscores, like `_PrivateTypeButPublicForReasons`. Do we want this rule to flag those too? Since those situations still come up in real-world code I don't know if we want to burden users with having to write `// swift-format-ignore: TypeNamesShouldBeCapitalized` above those.

Maybe strip off any sequence of leading underscores before checking the "first" character?

Reply:
Sounds good!

Reply:
You probably want to instead check that the first character *does not change when uppercased* to make it independent of the `IdentifiersMustBeASCII` rule.

Reply:
Addressed this is in 2fe98cf4ee5b018ffb5cac22eb8030e347e9f634